python/ray/serve/multiplex.py

```diff
@@ -0,0 +1,94 @@
+    be a coroutine function that takes the model ID as the first argument and
+    returns the user-constructed model object.
+    The model multiplexer will also ensure that the number of models on the current
+    replica does not exceed the specified limit.
+    The model will be unloaded in the LRU order, the model multiplexer will call the
+    model's __del__ attribute if it exists to clean up the model resources eagerly.
+
+    """
+
+    def __init__(
+        self,
+        model_load_func: Callable[[str], Any],
+        self_arg: Any,
+        max_num_models_per_replica: int,
+    ):
+        """Initialize the model multiplexer.
+        Args:
+            model_load_func: the model load async function.
+            self_arg: self argument when model_load_func is class method.
+            max_num_models_per_replica: the maximum number of models to be loaded on the
+                current replica. If it is -1, there is no limit for the number of models
+                per replica.
+        """
+        self.models = OrderedDict()
+        self._func = model_load_func
+        self.self_arg = self_arg
+        self.max_num_models_per_replica = max_num_models_per_replica
+
+    async def load_model(self, model_id: str) -> Any:
+        """Load the model if it is not loaded yet, and return the user-constructed model object.
+
+        Args:
+            model_id: the model ID.
+
+        Returns:
+            The user-constructed model object.
+        """
+
+        if type(model_id) != str:
+            raise TypeError("The model ID must be a string.")
+
+        if not model_id:
+            raise ValueError("The model ID cannot be empty.")
+
+        if model_id in self.models:
+            # Move the model to the end of the OrderedDict to ensure LRU caching.
+            model = self.models.pop(model_id)
+            self.models[model_id] = model
+        else:
+            # If the number of models per replica is specified, check if the number of
+            # models on the current replica has reached the limit.
+            if self.max_num_models_per_replica > 0:
+                if len(self.models) >= self.max_num_models_per_replica:
+                    # Unload the least recently used model.
+                    await self.unload_model()
+            # Load the model.
+            logger.info("Loading model '{}'.".format(model_id))
+            if self.self_arg is None:
+                self.models[model_id] = await self._func(model_id)
+            else:
+                self.models[model_id] = await self._func(self.self_arg, model_id)
+        return self.models[model_id]
+
+    async def unload_model(self) -> None:
+        """Unload the least recently used model."""
+        model_id, model = self.models.popitem(last=False)
+        logger.info("Unloading model '{}'.".format(model_id))
+
+        # If the model has __del__ attribute, call it.
+        # This is to clean up the model resources eagerly.
```

Comment:
I believe we need to overwrite the `__del__` attribute after doing this so that it isn't called twice (which might have some weird side effects)